pints/_nested/__init__.py

```diff
@@ -50,12 +230,424 @@ def __init__(self, log_likelihood, log_prior):
+            self._logger.add_counter('Eval.', max_value=self._iterations * 10)
+            #TODO: Add other informative fields ?
+            self._logger.add_time('Time m:s')
+            self._logger.add_float('Delta_log(z)')
+            self._logger.add_float('Acceptance rate')
+
+    def _initial_points(self):
+        """
+        Generates initial active points.
+        """
+        m_initial = self._log_prior.sample(self._n_active_points)
+        v_fx = np.zeros(self._n_active_points)
+        for i in range(0, self._n_active_points):
+            # Calculate likelihood
+            v_fx[i] = self._evaluator.evaluate([m_initial[i, :]])[0]
+            self._sampler._n_evals += 1
+
+            # Show progress
+            if self._logging and i >= self._next_message:
+                # Log state
+                self._logger.log(0, self._sampler._n_evals,
+                                 self._timer.time(), self._diff, 1.0)
+
+                # Choose next logging point
+                if i > self._message_warm_up:
+                    self._next_message = self._message_interval * (
+                        1 + i // self._message_interval)
+        return v_fx, m_initial
+
+    def _diff_marginal_likelihood(self, i, d):
+        """
+        Calculates difference in marginal likelihood between current and
+        previous iterations.
+        """
+        v_temp = np.concatenate((
+            self._v_log_Z[0:(i - 1)],
+            [np.max(self._sampler._m_active[:, d])]))
+        w_temp = np.concatenate((self._w[0:(i - 1)], [self._X[i]]))
+        self._diff = (logsumexp(self._v_log_Z[0:(i - 1)],
+                                b=self._w[0:(i - 1)]) -
+                      logsumexp(v_temp, b=w_temp))
+
+    def _update_logger(self):
+        """
+        Updates logger if necessary.
+        """
+        if self._logging:
+            self._i_message += 1
+            if self._i_message >= self._next_message:
+                # Log state
+                self._logger.log(self._i_message, self._sampler._n_evals,
+                                 self._timer.time(), self._diff,
+                                 float(self._sampler._accept_count /
+                                       (self._sampler._n_evals -
+                                        self._sampler._n_active_points)))
+
+                # Choose next logging point
+                if self._i_message > self._message_warm_up:
+                    self._next_message = self._message_interval * (
+                        1 + self._i_message // self._message_interval)
+
     def run(self):
         """
         Runs the nested sampling routine and returns a tuple of the posterior
         samples and an estimate of the marginal likelihood.
         """
-        raise NotImplementedError
+
+        # Choose method to evaluate
+        f = self._initialise_sensitivities()
+
+        # Set parallel
+        self._evaluator = self._initialise_parallel(f)
```

Comment:
There's an awful lot of `self._x` going on here. Are they all necessary? Might be a lot more readable if you pass in required arguments to these private methods, and return whatever it is they're meant to initialise.

Reply:
I did originally write it like that but the number of arguments became quite extensive (and easy to mix up), so would prefer to keep it this way, if that'd be ok?